conn_http.go

```diff
@@ -30,102 +35,178 @@ import (
+		return true
+	}
+	return false
+}
+
+func writeData(encoder *binary.Encoder, block *proto.Block) error {
+	return block.Encode(encoder, 0)
+}
+
+func readData(decoder *binary.Decoder) (*proto.Block, error) {
+	var block proto.Block
+	if err := block.Decode(decoder, 0); err != nil {
+		return nil, err
+	}
+	return &block, nil
+}
+
+func (h *httpConnect) asyncInsert(ctx context.Context, query string, wait bool) error {
+	return errors.New("HTTP: not supported")
 }
 
-func (h *httpConnect) prepareRequest(ctx context.Context, query string, args ...interface{}) (*http.Request, error) {
-	query, err := bind(h.location, query, args)
+func readResponse(response *http.Response) ([]byte, error) {
+	var result []byte
+	if response.ContentLength > 0 {
+		result = make([]byte, 0, response.ContentLength)
+	}
+	buf := bytes.NewBuffer(result)
+	defer response.Body.Close()
+	_, err := buf.ReadFrom(response.Body)
+
 	if err != nil {
 		return nil, err
 	}
 
-	reader := strings.NewReader(query)
+	return buf.Bytes(), nil
+}
+
+func (h *httpConnect) prepareRequest(ctx context.Context, reader io.Reader, extra map[string]string) (*http.Request, error) {
 
-	req, err := http.NewRequest(http.MethodPost, h.url.String(), reader)
+	u := *h.url
+
+	if len(extra) != 0 {
+		query := u.Query()
+		for key, value := range extra {
+			query.Set(key, value)
+		}
+		u.RawQuery = query.Encode()
+	}
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u.String(), reader)
 
 	return req, err
 }
 
 func (h *httpConnect) executeRequest(ctx context.Context, req *http.Request) (io.ReadCloser, error) {
 
-	if h.transport == nil {
+	if h.client == nil {
 		return nil, driver.ErrBadConn
 	}
 
-	resp, err := h.transport.RoundTrip(req)
+	resp, err := h.client.Do(req)
 	if err != nil {
 		return nil, err
 	}
 	if resp.StatusCode != http.StatusOK {
-		return nil, fmt.Errorf("clickhouse [execute]:: got no 200 code('%d')", resp.StatusCode)
+		msg, err := readResponse(resp)
+
+		if err != nil {
+			return nil, errors.Wrap(err, "clickhouse [execute]:: failed to read the response")
+		}
+
+		return nil, fmt.Errorf("clickhouse [execute]:: %d code: %s", resp.StatusCode, string(msg))
 	}
 
 	return resp.Body, nil
 }
 
-func (h *httpConnect) ping(ctx context.Context) error {
-	req, err := h.prepareRequest(ctx, "SELECT 1")
+func (h *httpConnect) exec(ctx context.Context, query string, args ...interface{}) error {
+	query, err := bind(h.location, query, args...)
 	if err != nil {
 		return err
 	}
 
-	res, err := h.executeRequest(ctx, req)
+	req, err := h.prepareRequest(ctx, strings.NewReader(query), nil)
 	if err != nil {
 		return err
 	}
-	s, err := io.ReadAll(res)
+
+	res, err := h.executeRequest(ctx, req)
+	if res != nil {
+		defer res.Close()
+		// we don't care about result, so just discard it to reuse connection
+		_, _ = io.Copy(ioutil.Discard, res)
+	}
+
+	return err
+}
+
+func (h *httpConnect) ping(ctx context.Context) error {
+	rows, err := h.query(ctx, nil, "SELECT 1")
```

Comment:
there is a ping endpoint for http i.e. "In health-check scripts use GET /ping request. This handler always returns “Ok.” (with a line feed at the end)."

Reply:
I dont think this is that important btw, optional change